daml/Tests/Trading/E2E.daml

```diff
@@ -53,71 +48,215 @@ tradingEndToEndTest = do
+
+  (aliceOrderCid, aliceOrderStateCid) <- submit ccp do
+    exerciseCmd aliceOrderReqCid Trading.AcknowledgeRequest
+
+  let bobOrderId = "bobExchangeOrder"
+  bobOrderReqCid <- submit bob.customer do
+    exerciseCmd bobCcpExecutionServiceCid $ Trading.RequestCreateOrder
+      Order.Order with
+        originator = bob.customer
+        executor = ccp
+        id = bobOrderId
+        listingId
+        asset = shareAsset
+        side = Order.Buy
+        timeInForce = Order.GTC
+        price = 100.0
+
+  -- FIXME: novate in a single step, so that order is filled and Ack'd at the same time
+  (bobOrderCid, bobOrderStateCid) <- submit ccp do 
+    exerciseCmd bobOrderReqCid $ Trading.AcknowledgeRequest
+  
+  Some aliceOrder <- queryContractId ccp aliceOrderCid
+  Some bobOrder <- queryContractId ccp bobOrderCid
+
+  ((_, _, aliceOrderStateCid, _), (_, _, bobOrderStateCid, _)) <- submit ccp do
+    b <- exerciseCmd bobOrderCid $ Order.Fill { counterparty = ccp, contraOrder = aliceOrder {originator = ccp , executor = bob.customer} }
+    a <- exerciseCmd aliceOrderCid $ Order.Fill { counterparty = ccp, contraOrder = bobOrder {originator = ccp, executor = alice.customer} }
+    return (a, b)
+  
+  Some Order.State{status=Order.FullyExecuted} <- queryContractId alice.customer aliceOrderStateCid
+  Some Order.State{status=Order.FullyExecuted} <- queryContractId bob.customer bobOrderStateCid
+  
+  -- Scenario # 4 Inter-dealer broker
+  -- ############################################################################
+
+  -- We're re-using the services created above for this
+
+  let aliceOrderId = "aliceAgencyOrder"
+  aliceOrderReqCid <- submit alice.customer do
+    exerciseCmd bobExecutionServiceCid $ Trading.RequestCreateOrder
+      Order.Order with
+        originator = alice.customer
+        executor = bob.customer
+        id = aliceOrderId
+        listingId
+        asset = shareAsset
+        side = Order.Sell
+        timeInForce = Order.GTC
+        price = 100.0
+  (aliceOrderCid, aliceOrderStateCid) <- submit bob.customer do
+    exerciseCmd aliceOrderReqCid Trading.AcknowledgeRequest
+
+  Some Order.State{status=Order.PendingExecution} <- 
+    queryContractId alice.customer aliceOrderStateCid
+
+  Some aliceOrder <- queryContractId bob.customer aliceOrderCid
+
+  let bobOrderId = "bobAgencyOrder"
+  bobOrderReqCid <- submit bob.customer do
+    exerciseCmd bobCcpExecutionServiceCid $ Trading.RequestCreateOrder
+      aliceOrder with originator = bob.customer, executor = ccp, id = bobOrderId
+
+  (bobOrderCid, bobOrderStateCid) <- submit ccp do 
+     exerciseCmd bobOrderReqCid $ Trading.AcknowledgeRequest
 
-  aliceOrderCid <- fst . head <$> query @Order.T alice.customer
-  checkOrderStatusMatches aliceOrderCid alice.customer Order.FullyExecuted
-  bobOrderCid <- fst . head <$> query @Order.T bob.customer
-  checkOrderStatusMatches bobOrderCid bob.customer Order.FullyExecuted
+  -- FIXME : make the two fills atomic
```

Comment:
Here, as well as novation above, it would be useful to be able to do these transactions in a single transaction. Look at adding specific choices for that.